pkg/controllers/inflightchecks/controller.go

```diff
@@ -0,0 +1,121 @@
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/utils/clock"
+	"knative.dev/pkg/logging"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/aws/karpenter-core/pkg/apis/provisioning/v1alpha5"
+	"github.com/aws/karpenter-core/pkg/cloudprovider"
+	"github.com/aws/karpenter-core/pkg/controllers/deprovisioning"
+	"github.com/aws/karpenter-core/pkg/controllers/state"
+	"github.com/aws/karpenter-core/pkg/events"
+	operatorcontroller "github.com/aws/karpenter-core/pkg/operator/controller"
+	"github.com/aws/karpenter-core/pkg/operator/injection"
+)
+
+const controllerName = "inflightchecks"
+
+type Controller struct {
+	clock      clock.Clock
+	kubeClient client.Client
+	checks     []Check
+	recorder   events.Recorder
+}
+type Check interface {
+	// Check performs the inflight check, this should return a list of slice discovered, or an empty
+	// slice if no issues were found
+	Check(ctx context.Context, node *v1.Node, provisioner *v1alpha5.Provisioner, pdbs *deprovisioning.PDBLimits) ([]Issue, error)
+}
+
+type Issue struct {
+	Node    *v1.Node
+	Message string
+}
+
+func NewController(clk clock.Clock, kubeclient client.Client, recorder events.Recorder, provider cloudprovider.CloudProvider, cluster *state.Cluster) *Controller {
+	return &Controller{
+		clock:      clk,
+		kubeClient: kubeclient,
+		recorder:   recorder,
+		checks: []Check{
+			NewFailedInit(clk, provider),
+			NewTermination(kubeclient),
+			NewNodeShape(provider),
+		}}
+}
+
+func (c *Controller) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	log := logging.FromContext(ctx).Named(controllerName).With("node", req.Name)
+	ctx = logging.WithLogger(ctx, log)
+	ctx = injection.WithControllerName(ctx, controllerName)
+
+	node := &v1.Node{}
+	if err := c.kubeClient.Get(ctx, req.NamespacedName, node); err != nil {
+		if errors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	provisioner := &v1alpha5.Provisioner{}
+	if err := c.kubeClient.Get(ctx, types.NamespacedName{Name: node.Labels[v1alpha5.ProvisionerNameLabelKey]}, provisioner); err != nil {
+		// provisioner is missing, node should be removed soon
+		return reconcile.Result{}, client.IgnoreNotFound(err)
+	}
+
+	pdbs, err := deprovisioning.NewPDBLimits(ctx, c.kubeClient)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	var allIssues []Issue
+	for _, check := range c.checks {
+		issues, err := check.Check(ctx, node, provisioner, pdbs)
+		if err != nil {
+			log.Errorf("checking node %s with %T, %s", node.Name, check, err)
+		}
+		allIssues = append(allIssues, issues...)
+	}
+	for _, iss := range allIssues {
+		log.Infof("Inflight check failed for node %s, %s", iss.Node.Name, iss.Message)
```

Comment:
Possible to encapsulate in the recorder?

Reply:
It would need to check for event types there, we don't want to log on all event types.  There is a hack in the code currently to prevent controller-runtime from logging all events as they are too numerous. Because of that, we should log them here.